rfcs/0075-declarative-wrappers.md

```diff
@@ -0,0 +1,372 @@
+    "/nix/...gsettings-desktop-schemas.../..."
+  ],
+  "GDK_PIXBUF_MODULE_FILE": "/nix/...librsvg.../lib/gdk.../loaders.cache",
+}
+```
+
+The information found inside an input's `wrappers.json` will specify the
+wrapper information not only for itself, but for all its dependencies (including transitives) as well. Thus, in
+contrast to the [POC Nixpkgs PR](https://github.com/NixOS/nixpkgs/pull/85103)
+and the [original design of the
+RFC](https://github.com/doronbehar/rfcs/blob/60d3825fdd4e6574b7e5d70264445d1c801368c6/rfcs/0075-declarative-wrappers.md#L251),
+prior to [the 1st
+meeting](https://github.com/NixOS/rfcs/pull/75#issuecomment-760942876),
+traversing all the inputs and the inputs' inputs, will not happen during eval
+time and only partly, during build time - every package already built will
+provide its reverse dependencies all the information they need about
+environment variables.
+
+Most of the work to do will be:
+
+1. Gather information about what environment variables are "linked" to each
+   package, and edit these derivations to include a `wrappers.json` in them.
+   This should be done with `makeWrapperAuto` as well, see (2).
+2. Design the `makeWrapperAuto` shell hook:
+  - It should introduce a shell function (to be called `combineWrappersInfo`) that
+    will allow piping a JSON string from `builtins.toJSON` and spit a
+    `wrappers.json` that will include both what was piped into it, and the
+    content from the package's various inputs' `wrappers.json` files.
+  - It should wrap the executables in `$out/bin/` according to
+    what's currently in this package's `wrappers.json`, during `fixupPhase`.
+  - The above should be also possible to do manually for executables outside
+    `$out/bin/` by setting `wrapExtraPrograms` on the derivation:
+
+```nix
+  wrapExtraPrograms = [ "/libexec/" "/share/scripts" ];
+```
+
+3. Most of the packages with linked environment variables, have lots of reverse
+   dependencies, so once `makeWrapperAuto` is ready, it'd nice to have a hydra
+   job that will build all of these packages with the `wrappers.json` file in
+   them. For instance these packages include:
+   - `gdk-pixbuf`
+   - `gsettings-desktop-schemas`
+   - `pango`
+   - `gtk3`
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+When switching to `makeWrapperAuto` from `makeWrapper` there shouldn't be
+manual usage of `wrapProgram` for most cases. A package that uses `wrapProgram`
+should be able to switch to `combineWrappersInfo` and declare any nontrivial
+environment variables with it to get propagated to reverse dependencies and to
+it's executables automatically.
+
+Currently I imagine the usage of `combineWrappersInfo` (the name can be debated) as
+so:
+
+```nix
+  # Propagate GST plugins' path
+  postInstall = ''
+    echo "${builtins.toJSON {
+      GST_PLUGIN_SYSTEM_PATH_1_0 = [
+        # @out@ should be expanded by `combineWrappersInfo` to what's in `$out`, see:
```

Comment:
We should probably use [`${placeholder "out"}`](https://nixos.org/manual/nix/stable/#builtin-placeholder) here? Or am I missing something?

Reply:
> We should probably use [`${placeholder "out"}`](https://nixos.org/manual/nix/stable/#builtin-placeholder) here? Or am I missing something?

A lot of people ask that, see https://github.com/NixOS/nixpkgs/pull/85103#issuecomment-613071343 - I linked that explanation in the comment in the code snippet.